zerver/views/auth.py

```diff
@@ -543,8 +555,73 @@ def update_login_page_context(request: HttpRequest, context: Dict[str, Any]) ->
 
     context['wrong_subdomain_error'] = WRONG_SUBDOMAIN_ERROR
 
+class PatchedExtraSessionStorage(ExtraSessionStorage):
+    def reset(self) -> None:
+        if self.prefix in self.request.session:
+            super(ExtraSessionStorage, self).reset()
+        else:
+            self.init_data()
+
+class TwoFactorLoginView(BaseTwoFactorLoginView):
+    extra_context = None  # type: ExtraContext
+    storage_name = 'zerver.views.auth.PatchedExtraSessionStorage'
+    form_list = (
+        ('auth', OurAuthenticationForm),
+        ('token', AuthenticationTokenForm),
+        ('backup', BackupTokenForm),
+    )
+
+    def __init__(self, extra_context: ExtraContext=None,
+                 *args: Any, **kwargs: Any) -> None:
+        self.extra_context = extra_context
+        super().__init__(*args, **kwargs)
+
+    def get_form_kwargs(self, step: str) -> Dict[str, Any]:
+        kwargs = super(TwoFactorLoginView, self).get_form_kwargs(step)
+        if step == 'auth':
+            kwargs.update({'request': self.request})
+        return kwargs
+
+    def get_context_data(self, **kwargs: Any) -> Dict[str, Any]:
+        context = super(TwoFactorLoginView, self).get_context_data(**kwargs)
+        if self.extra_context is not None:
+            context.update(self.extra_context)
+        update_login_page_context(self.request, context)
+
+        realm = get_realm_from_request(self.request)
+        redirect_to = realm.uri if realm else '/'
+        context['next'] = self.request.GET.get('next', redirect_to)
+        return context
+
+    def done(self, form_list: List[Form], **kwargs: Any) -> HttpResponse:
+        """
+        Login the user and redirect to the desired page.
+
+        We need to override this function so that we can redirect to
+        realm.uri instead of '/'.
```

Comment:
Just to confirm I read this correctly, this function is equivalent to this hacky pseudocode, right?:

`with self.settings(LOGIN_REDIRECT_URL=self.get_user().realm.uri): super().done()`, right?